drivers/SPI.cpp

```diff
@@ -16,97 +16,144 @@
 
-#if DEVICE_SPI_ASYNCH && TRANSACTION_QUEUE_SIZE_SPI
+#if 0 && TRANSACTION_QUEUE_SIZE_SPI
 CircularBuffer<Transaction<SPI>, TRANSACTION_QUEUE_SIZE_SPI> SPI::_transaction_buffer;
 #endif
+SPI::spi_peripheral_s SPI::_peripherals[SPI_COUNT];
 
 SPI::SPI(PinName mosi, PinName miso, PinName sclk, PinName ssel) :
-    _spi(),
-#if DEVICE_SPI_ASYNCH
-    _irq(this),
+    _self(),
+#if 0
     _usage(DMA_USAGE_NEVER),
     _deep_sleep_locked(false),
 #endif
     _bits(8),
-    _mode(0),
+    _mode(SPI_MODE_IDLE_LOW_SAMPLE_FIRST_EDGE),
+    _msb_first(true),
     _hz(1000000),
-    _write_fill(SPI_FILL_CHAR)
+    _write_fill(0xFF)
 {
     // No lock needed in the constructor
+    // get_module
 
-    spi_init(&_spi, mosi, miso, sclk, ssel);
-    _acquire();
+    SPIName name = spi_get_module(mosi, miso, sclk);
+
+    // this is a lazy allocation for the mutex.
+    // unless not assigned to a peripheral, this mutex will never be deleted.
+    PlatformMutex *mutex = new PlatformMutex();
+
+    core_util_critical_section_enter();
+    // lookup in a critical section if we already have it else initialize it
+    _self = SPI::lookup(name, true);
+    if (_self->mutex == NULL) {
+        _self->mutex = mutex;
+    }
+    if (_self->name == 0) {
+        _self->name = name;
+        spi_init(&_self->spi, false, mosi, miso, sclk, NC);
+    }
+    core_util_critical_section_exit();
+
+    if (_self->mutex != mutex) {
+        delete mutex;
+    }
+
+    // we don't need to _acquire at this stage.
+    // this will be done anyway before any operation.
+}
+
+struct SPI::spi_peripheral_s *SPI::lookup(SPIName name, bool or_last) {
+    struct SPI::spi_peripheral_s *result = NULL;
+    core_util_critical_section_enter();
+    for (uint32_t idx = 0; idx < SPI_COUNT; idx++) {
+        if ((_peripherals[idx].name == name) ||
+            (_peripherals[idx].name == 0)) {
+            result = &_peripherals[idx];
+            break;
+        }
+        // XXX: we may want to ensure that it was previously initialized with the same mosi/miso/sclk/ss pins
+    }
+    if (!or_last && (result != NULL) && (result->name == 0)) {
+        result = NULL;
+    }
+
+    core_util_critical_section_exit();
+    return result;
 }
 
 void SPI::format(int bits, int mode)
+{
+    format(bits, (spi_mode_t)mode, true);
+}
+
+void SPI::format(uint8_t bits, spi_mode_t mode, bool msb_first)
 {
     lock();
     _bits = bits;
     _mode = mode;
+    _msb_first = msb_first;
     // If changing format while you are the owner then just
     // update format, but if owner is changed then even frequency should be
     // updated which is done by acquire.
-    if (_owner == this) {
-        spi_format(&_spi, _bits, _mode, 0);
+    if (_self->owner == this) {
+        spi_format(&_self->spi, _bits, _mode, _msb_first?SPI_BIT_ORDERING_MSB_FIRST:SPI_BIT_ORDERING_LSB_FIRST);
     } else {
         _acquire();
     }
     unlock();
 }
 
-void SPI::frequency(int hz)
+uint32_t SPI::frequency(uint32_t hz)
```

Comment:
Variable/Parameter types should be consistent in driver. Existing driver uses `int`, `char`.. etc, changed here for few API's to `uint32_t` `uint8_t` and not all.  I would expect API's to be consistent in a driver. 

Also if we are changing the return type, should we deprecate the existing one?  In case of `format` for addition of new parameter we are adding new API, but old is not deprecated and for `frequency` we are changing the return type and parameter type of API without any deprecation.

Reply:
The changes to this driver are provisional and meant to be minimal.
Any change to bring consistency may break existing API.

We cannot deprecate the void returning version of `SPI::frequency` because return types are not "overloadable". Also, going from `void` to `uint32_t` do not present risk on ignoring this return value.